### ISSUE TITLE
Opaque - reference token validation

### ISSUE THREAD

**Issue by Simkiw** [dotnet/aspnetcore#46026]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe the problem.

Hi,

I have an identity provider OIDC compliant which issues a JWT ID token and an opaque access token (or reference token).
Microsoft offers a way to validate a JWT bearer however, I did not see an alternative for opaque tokens.

### Describe the solution you'd like

I would like to have a method extension that works the same as `ServiceCollection.AddJwtBearer()` extension.  

We would provide almost the same identity provider information such as the issuer uri, user info, and most importantly the **introspect endpoint**, which checks if the access token is valid or not.  

Finally, we could populate the identity principal with the userinfo endpoint, right after validating the token.

**Comment by blowdart** [comment]
The refresh token is only understood by the identity provider. They are the only service that can understand it or validate it.

It shouldn't be sent it to your api endpoints. It's for the client app to send to the identity issuer when your client gets an auth error. 

We can't add validation because we didn't issue it, and we don't know how the identity provider works, or understand the refresh token.

**Comment by kevinchalet** [comment]
@blowdart `reference tokens` != `refresh tokens`.

@Simkiw FWIW, OpenIddict has a validation handler that natively supports OAuth 2.0 introspection, which should work with any server implementation as long as it's standard-compliant. You can see https://github.com/openiddict/openiddict-samples/blob/dev/samples/Zirku/Zirku.Api1/Program.cs for an example.

**Comment by blowdart** [comment]
Aggh doh. This is what I get for opening github when I wake up :)

**Comment by blowdart** [comment]
@Simkiw does a third party option work for you, or are you hampered by corporate rules etc making it hard to use 3rd party code?

**Comment by Simkiw** [comment]
So far, I have freedom to use whatever i want.    
I also came across [this article](https://leastprivilege.com/2020/07/06/flexible-access-token-validation-in-asp-net-core/) written in 2020 which targets the same issue, but involves some code tweeking.  

As for the openiddict, no i wasn't aware of it. I'll take a look at it and give it a try.  

Thank you all, I'll keep you posted

**Comment by captainsafia** [comment]
Triage: this is a new feature that we will probably want to support via a callback API of some sort.

**Comment by Simkiw** [comment]
@wrschneider Hi there,  

Like you said, both userinfo and introspect serve different purposes.  
Usually if not always, you cannot get userinfo without a valid token, which means, if you get the userinfo details, that implies the token is valid.  

However, it is vital to keep in mind that the introspection endpoint requires authentication as opposed to userinfo endpoint :  
- Anyone could get the userinfo details as long as the http request has a valid token
- Only authenticated requests could check the validity of a token (as if you are sending two tokens, one authenticating the caller and the other one the token you are checking over)  

If somehow your access token gets compromised, userinfo endpoint might not be enough to secure the backend.

**Comment by wrschneider** [comment]
> If somehow your access token gets compromised, userinfo endpoint might not be enough to secure the backend.

This is why access tokens tend to be short-lived: If your access token is compromised, then you can call whatever API the access token authorizes you to call.  Userinfo is just one specific example.

No different than using `AddJwtBearer` to authenticate and using public keys to check JWT signature, or if you had some other kind of access token to authorize your back end API explicitly.

**Comment by iamjplant** [comment]
Agreed!  This would be nice to have for those of us who don't get JWT's back!

**Comment by GNUGradyn** [comment]
This is a massive shame. There are some inherit security compromises in using JWTs that are NOT worth the reduced database pressure for some use cases. This should be supported. Should have been supported a long time ago.
